docs/tools.md

```diff
@@ -0,0 +1,328 @@
+
+  * **Mechanism:** An agent obtains an instance of a `ToolBox` (or a utility function that wraps the process) and calls the public method `call_tool`.
+  * **Benefits:**
+      * **Direct Access:** No messagebus latency.
+      * **Strong Validation:** Enforces the `argument_schema` and `output_schema` before and after execution, ensuring data integrity.
+  * **Usage Flow (Internal to Agent):**
+    1.  Get the `ToolBox` instance (e.g., via a plugin loading utility).
+    2.  Call: `validated_output = toolbox.call_tool(name="tool_name", tool_kwargs={"arg1": value})`
+    3.  The return value is a fully validated **`ToolOutput`** Pydantic object.
+
+### 2. MessageBus Discovery and Calling
+
+This method is crucial for **cross-process communication** and **runtime discovery**, allowing any component with access to the OVOS MessageBus to find and execute tools.
+
+#### A. Tool Discovery
+
+Agents can broadcast a discovery message to find all available tools across all running ToolBoxes.
+
+  * **Discovery Message:**
+    ```json
+    {"type": "ovos.persona.tools.discover"}
+    ```
+  * **ToolBox Response:** Each registered `ToolBox` responds with its tools, providing a complete description that includes the Pydantic schemas converted to **JSON Schema** (via the `tool_json_list` property). This format is ideal for submitting to LLMs (e.g., for **Function Calling** or **Tool Use**).
+  * **Tool Schema Format (Simplified):**
+    ```json
+    {
+        "name": "web_search",
+        "description": "Searches the web for information...",
+        "argument_schema": {
+            "type": "object",
+            "properties": {
+                "query": {"type": "string", "description": "The search query."}
+            },
+            "required": ["query"]
+        },
+        // ... (output_schema also as JSON Schema)
+    }
+    ```
+
+#### B. Tool Execution (Remote Call)
+
+To execute a tool remotely, an agent sends a message targeted at the specific `ToolBox`.
+
+  * **Tool Call Message:**
+    ```json
+    {
+        "type": "ovos.persona.tools.<toolbox_id>.call",
+        "data": {
+            "name": "tool_name",
+            "kwargs": {"arg1": "value1", "arg2": 123} // Raw argument dict
+        }
+    }
+    ```
+  * **ToolBox Execution:** The target `ToolBox` receives the message, runs the **`call_tool`** logic (which includes validation), and emits a response message.
+  * **ToolBox Response Message (Success):**
+    ```json
+    {
+        "type": "ovos.persona.tools.<toolbox_id>.call.response",
+        "data": {
+            "toolbox_id": "web_search_tools",
+            "result": {"field1": "output_value", "field2": 456} // Raw dictionary output
+        }
+    }
+    ```
+  * **ToolBox Response Message (Error):**
+    ```json
+    {
+        "type": "ovos.persona.tools.<toolbox_id>.call.response",
+        "data": {
+            "toolbox_id": "web_search_tools",
+            "error": "ValueError: Invalid input for 'tool_name'..."
```

Comment:
`ToolOutput` is supposed to be the guaranteed response per the tool contract, but this would be a different format, wouldn't it? Is there going to be something like a `ToolError` class and a tool could return `ToolOutput | ToolError`?

Reply:
this is not the tool output itself, this is the messagebus handler, when an exception happened using the tool, so in this case there's no tool output because it failed